### ISSUE TITLE
Locks have a race condition

### ISSUE THREAD

**Issue by ewjoachim** [peopledoc/procrastinate#103]
Ok, we have, I think, our first real bug.

## TL;DR

If I launch `procrastinate_fetch_job` quickly twice, I can get 2 jobs that share the same `lock`.

## Proof

```console
$ psql -c 'SELECT * FROM procrastinate_jobs;'
 id | queue_name |           task_name            | lock |   args    | status | scheduled_at | started_at | attempts
----+------------+--------------------------------+------+-----------+--------+--------------+------------+----------
  1 | sleep      | procrastinate_demo.tasks.sleep | yay  | {"i": 20} | todo   |              |            |        0
  2 | sleep      | procrastinate_demo.tasks.sleep | yay  | {"i": 20} | todo   |              |            |        0
(2 rows)

$ psql -c 'SELECT procrastinate_fetch_job(NULL);' & psql -c 'SELECT procrastinate_fetch_job(NULL);'&
[1] 22792
[2] 22793
                                       procrastinate_fetch_job
-----------------------------------------------------------------------------------------------------
 (1,sleep,procrastinate_demo.tasks.sleep,yay,"{""i"": 20}",doing,,"2019-10-27 20:14:12.230384+00",0)
(1 row)

[2]  + 22793 done       psql -c 'SELECT procrastinate_fetch_job(NULL);'
                                       procrastinate_fetch_job
-----------------------------------------------------------------------------------------------------
 (2,sleep,procrastinate_demo.tasks.sleep,yay,"{""i"": 20}",doing,,"2019-10-27 20:14:12.233998+00",0)
(1 row)

[1]  + 22792 done       psql -c 'SELECT procrastinate_fetch_job(NULL);'
```
The second call should have returned `(,,,,,,,,)`.

For the sake of it, if I put the 2 calls in the same transaction, I DO have the expected result:

```console
$ psql -c 'SELECT * FROM procrastinate_jobs;'
 id | queue_name |           task_name            | lock |   args    | status | scheduled_at | started_at | attempts
----+------------+--------------------------------+------+-----------+--------+--------------+------------+----------
  1 | sleep      | procrastinate_demo.tasks.sleep | yay  | {"i": 20} | todo   |              |            |        0
  2 | sleep      | procrastinate_demo.tasks.sleep | yay  | {"i": 20} | todo   |              |            |        0
(2 rows)

$ psql -c 'SELECT procrastinate_fetch_job(NULL), procrastinate_fetch_job(NULL);'
                                      procrastinate_fetch_job                                       | procrastinate_fetch_job
----------------------------------------------------------------------------------------------------+-------------------------
 (1,sleep,procrastinate_demo.tasks.sleep,yay,"{""i"": 20}",doing,,"2019-10-27 20:17:46.28228+00",0) | (,,,,,,,,)
(1 row)
```

## Leads

- When we call [`FOR UPDATE OF procrastinate_jobs`](https://github.com/peopledoc/procrastinate/blob/master/procrastinate/sql/structure.sql#L58), should we somehow also lock `procrastinate_job_locks`? But there's no row to lock yet...
- Should there be a `UNIQUE` constraint on `procrastinate_job_locks.object`? I guess it would help but if we add just that, we'll get a crash, and still not the expected behaviour. Or should we also add a `ON CONFLICT DO NOTHING` statement? The procedure would probably need quite a refactor, because in this case it needs to **not** do `UPDATE procrastinate_jobs SET status = 'doing' **and** return nothing.
- Is it because the `procrastinate_job_locks` is `UNLOGGED`?

**Comment by ewjoachim** [comment]
OK, so by adding `UNIQUE` I get the contraint checked, and thus an exception. Now to see if I can transform the query so that if there's a constraint error, I don't do the update.
